scripts/ocsf-schemas.py

```diff
@@ -0,0 +1,173 @@
+
+
+class Writer:
+    def __init__(self, file: typing.TextIO):
+        self.file = file
+        self.indent = 0
+
+    def print(self, *args, **kwargs) -> None:
+        print(self.indent * " ", file=self.file, end="")
+        print(*args, **kwargs, file=self.file)
+
+    def comment(self, text: str) -> None:
+        width = COLUMN_LIMIT - self.indent - len("// ")
+        lines = textwrap.wrap(text, width)
+        for line in lines:
+            self.print("//", line)
+
+    def begin(self, *args, **kwargs) -> None:
+        self.print(*args, **kwargs)
+        self.indent += 2
+
+    def end(self, *args, **kwargs) -> None:
+        self.indent -= 2
+        if self.indent < 0:
+            raise ValueError
+        self.print(*args, **kwargs)
+
+
+def _emit(writer: Writer, schema: Schema, *, objects: bool) -> None:
+    name = "objects" if objects else "classes"
+    prefix = OBJECT_PREFIX if objects else CLASS_PREFIX
+    types = schema["types"]
+    first = True
+    for entity in schema[name].values():
+        if not first:
+            writer.print()
+        first = False
+        if DOCUMENT_ENTITIES:
+            writer.comment(entity["description"])
+        entity_name = entity["name"]
+        omit = OMIT.get(entity_name, [])
+        writer.begin(f"type {prefix}{entity_name} = record{{")
+        for attr_name, attr_def in sorted(entity["attributes"].items()):
+            if attr_name in omit:
+                continue
+            if "profile" in attr_def:
+                continue
+            if "object_type" in attr_def:
+                type_name = attr_def["object_type"]
+                # Some object names have a prefix, which is separated by a
+                # slash. It looks like this prefix can just be discarded.
+                slash = type_name.rfind("/")
+                if slash != -1:
+                    type_name = type_name[slash + 1 :]
+                resolved = f"{OBJECT_PREFIX}{type_name}"
+            else:
+                resolved = types[attr_def["type"]]
+                if resolved is OMIT_MARKER:
+                    continue
+            if attr_def.get("is_array", False):
+                resolved = f"list<{resolved}>"
+            if DOCUMENT_FIELDS:
+                writer.comment(attr_def["description"])
+            writer.print(f"{attr_name}: {resolved},")
+        writer.end("}")
+
+
+def emit_classes(writer: Writer, schema: Schema) -> None:
+    log("Emitting class definitions")
+    _emit(writer, schema, objects=False)
+
+
+def emit_objects(writer: Writer, schema: Schema) -> None:
+    log("Emitting object definitions")
+    _emit(writer, schema, objects=True)
+
+
+def open_schema_file():
+    scripts = Path(__file__).parent
+    types = (scripts / "../schema/types").resolve()
+    if not types.is_dir():
+        raise NotADirectoryError(f"expected {types} to be a directory")
+    path = types / "ocsf.schema"
+    log(f"Opening schema file {path}")
+    return path.open("w")
```

Comment:
I  don't quite understand why this path is hardcoded. Writing to `./ocsf.schema` would be more obvious and carry less assumptions.

Reply:
I'd assume you'd almost always want to write to that path, so this makes the common case easy and prevents running the script in the wrong directory without noticing. If you agree that this is good enough, please resolve. Otherwise I would add some option like `-o`, though I cannot see this being used too much.

Reply:
Also, other scripts do the same (for example `update-plugins.sh`).